pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -56,87 +69,32 @@ func (r *PullSecretReconciler) Reconcile(request ctrl.Request) (ctrl.Result, err
 	// TODO(mj): controller-runtime master fixes the need for this (https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L93) but it's not yet released.
 	ctx := context.Background()
 
-	mysec, err := r.kubernetescli.CoreV1().Secrets(operator.Namespace).Get(ctx, operator.SecretName, metav1.GetOptions{})
+	operatorSecret, err := r.kubernetescli.CoreV1().Secrets(operator.Namespace).Get(ctx, operator.SecretName, metav1.GetOptions{})
 	if err != nil {
 		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		ps, isCreate, err := r.pullsecret(ctx)
-		if err != nil {
-			return err
-		}
-
-		if ps.Data == nil {
-			ps.Data = map[string][]byte{}
-		}
-
-		// validate
-		if !json.Valid(ps.Data[corev1.DockerConfigJsonKey]) {
-			r.log.Info("pull secret is not valid json - recreating")
-			delete(ps.Data, corev1.DockerConfigJsonKey)
-		}
-
-		pullsec, changed, err := pullsecret.Merge(string(ps.Data[corev1.DockerConfigJsonKey]), string(mysec.Data[corev1.DockerConfigJsonKey]))
-		if err != nil {
-			return err
-		}
-
-		// repair Secret type
-		if ps.Type != corev1.SecretTypeDockerConfigJson {
-			ps = &corev1.Secret{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      pullSecretName.Name,
-					Namespace: pullSecretName.Namespace,
-				},
-				Type: corev1.SecretTypeDockerConfigJson,
-				Data: map[string][]byte{},
-			}
-			isCreate = true
-			r.log.Info("pull secret has the wrong type - recreating")
+	var userSecret *corev1.Secret
 
-			// unfortunately the type field is immutable.
-			err = r.kubernetescli.CoreV1().Secrets(ps.Namespace).Delete(ctx, ps.Name, metav1.DeleteOptions{})
-			if err != nil {
-				return err
-			}
-
-			// there is a small risk of crashing here: if that happens, we will
-			// restart, create a new pull secret, and will have dropped the rest
-			// of the customer's pull secret on the floor :-(
-		}
-		if !isCreate && !changed {
-			return nil
-		}
-
-		ps.Data[corev1.DockerConfigJsonKey] = []byte(pullsec)
+	userSecret, err = r.kubernetescli.CoreV1().Secrets(pullSecretName.Namespace).Get(ctx, pullSecretName.Name, metav1.GetOptions{})
+	if err != nil && !kerrors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
 
-		if isCreate {
-			r.log.Info("re-creating pull secret")
-			_, err = r.kubernetescli.CoreV1().Secrets(ps.Namespace).Create(ctx, ps, metav1.CreateOptions{})
-		} else {
-			r.log.Info("updating pull secret")
-			_, err = r.kubernetescli.CoreV1().Secrets(ps.Namespace).Update(ctx, ps, metav1.UpdateOptions{})
-		}
-		return err
+	// reconcile global pull secret
+	// detects if the global pull secret is broken and fixes it by using backup managed by ARO operator
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// fix pull secret if its broken to have at least the ARO pull secret
+		return r.fixAndUpdateGlobalPullSecret(ctx, operatorSecret, userSecret)
 	})
```

Comment:
this logic is wrong.  The whole point of RetryOnConflict is that if there's a conflict, it redoes the Get.  The get should be inside the RetryOnConflict.